### PR TITLE
[LLVM] convergence verifier should visit all instructions

### PR DESCRIPTION
**Pull request by ssahasra**
The entry and loop intrinsics for convergence control cannot be preceded by convergent operations in their respective basic blocks. To check that, the verifier needs to reset its state at the start of the block. This was missed in the previous commit fa6dd7a24af2b02f236ec3b980d9407e86c2c4aa.

### CODE REVIEW

File: llvm/lib/IR/Verifier.cpp [hunk 1]
```diff
@@ -4805,6 +4803,8 @@ void Verifier::visitInstruction(Instruction &I) {
   BasicBlock *BB = I.getParent();
   Check(BB, "Instruction not embedded in basic block!", &I);
 
+  CV.visit(I);
```
Comment (arsenm) [inline llvm/lib/IR/Verifier.cpp:4806]: CV should probably be renamed, I had no idea what this was

File: llvm/lib/IR/Verifier.cpp [hunk 2]
```diff
@@ -3576,8 +3576,6 @@ void Verifier::visitCallBase(CallBase &Call) {
   if (Call.isInlineAsm())
     verifyInlineAsmCall(Call);
 
-  CV.visit(Call);
```
Comment (arsenm) [inline llvm/lib/IR/Verifier.cpp:3579]: If only calls can be convergent, I'm not sure I see how this is different. I don't really expect a visit instruction for the verifier to be side effecting. Can you have a separate enter block visitor?

Comment (ruiling) [reply llvm/lib/IR/Verifier.cpp:3579]: The ConvergenceVerifier may also keep record of the last-call-instruction's parent. Then it is easy to detect entering a new block inside ConvergenceVerifier::visit(). I have no opinion which is the best way to way.

Comment (ssahasra) [reply llvm/lib/IR/Verifier.cpp]: Fixed this by adding a visitBlock() method to the convergence verifier.

File: llvm/test/Verifier/convergencectrl-invalid.ll [hunk 3]
```diff
@@ -126,13 +126,18 @@ define void @entry_in_convergent(i32 %x, i32 %y) {
...
+; CHECK-NEXT: %h1
+; CHECK-SAME: %t60_tok3
 define void @loop_at_start(i32 %x, i32 %y) convergent {
 A:
   %t60_tok3 = call token @llvm.experimental.convergence.entry()
   br label %B
 B:
   %z = add i32 %x, %y
+  ; This is not an error
+  %h2 = call token @llvm.experimental.convergence.loop() [ "convergencectrl"(token %t60_tok3) ]
```
Comment (nhaehnle) [inline llvm/test/Verifier/convergencectrl-invalid.ll:138]: Maybe make that explicit with a CHECK-NOT: %h2?

Comment (ssahasra) [reply llvm/test/Verifier/convergencectrl-invalid.ll:138]: That's what the "CHECK-NEXT: %h1" does. It ensures that the error message is for %h1 and not %h2. There is not enough context in the stream of error messages to put a CHECK-NOT here.

File: llvm/lib/IR/Verifier.cpp (hunk 1 above)
Comment (ssahasra) [reply llvm/lib/IR/Verifier.cpp]: Fixed that.

**Comment by ssahasra** [comment]
Bump! Waiting for this one to the next PR that introduces convergence control operations and LLVM::token() type in GlobalISel.

Will be restarting this review as a new PR:
https://reviews.llvm.org/D158147

**Review by jayfoad — APPROVED** [review]
LGTM

**Pull request merged**